src/metabase/api/collection.clj

```diff
@@ -541,16 +542,65 @@
   [_ collection options]
   (collection-query collection options))
 
+(defn- annotate-collections
+  [parent-coll colls]
+  (let [visible-collection-ids (collection/permissions-set->visible-collection-ids
+                                @api/*current-user-permissions-set*)
+
+        child-type->coll-id-set
+        (reduce (fn [acc {collection-id :collection_id, card-type :type, :as _card}]
+                  (update acc (if (= (keyword card-type) :model) :dataset :card) conj collection-id))
+                {:dataset #{}
+                 :card    #{}}
+                (mdb.query/reducible-query {:select-distinct [:collection_id :type]
+                                            :from            [:report_card]
+                                            :where           [:= :archived false]}))
+
+        collections-containing-dashboards
+        (->> (t2/query {:select-distinct [:collection_id]
+                        :from :report_dashboard
+                        :where [:= :archived false]})
+             (map :collection_id)
+             (into #{}))
+
+        descendants (collection/descendants-flat parent-coll (collection/visible-collection-ids->honeysql-filter-clause
```

Comment:
`descendants` is shadowing `clojure.core/descendants`, would be nice if we could avoid that